### PR TITLE
feat: oracle

### PR DESCRIPTION
**Pull request by sokel**
Add Oracle script;
This implementation works through MultiSig contract.
For prevent races it have two modes:
- In master mode script submitting price change to multisig
- In slave mode script listen submitting from master, verify price, and confirm it.

Any mode script have price watcher, for timely update SNM-USD price.
Watcher loads data from cmc SNM ticker.

chained with #1069

### CODE REVIEW

File: cmd/oracle/main.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to load Ethereum keys: %s", err)
+	}
+
+	o, err := oracle.NewOracle(ctx, key, cfg)
+	if err != nil {
+		return fmt.Errorf("failed to build Oracle instance: %s", err)
+	}
+
+	if err := o.Serve(); err != nil {
```
Comment (antmat) [inline cmd/oracle/main.go]: IMO better to pass context to Serve

File: cmd/oracle/main.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		return fmt.Errorf("termination: %s", err)
+	}
+
+	defer o.Close()
+
+	return nil
```
Comment (antmat) [inline cmd/oracle/main.go]: signal processing?

File: etc/oracle.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  pass_phrase: "111111111"
+
+blockchain:
+  # endpoint: "https://rinkeby.infura.io/00iTrs5PIy0uGODwcsrb"
+  # sidechain_endpoint: "https://sidechain-dev.sonm.com"
+
+oracle:
+  # In master mode this instance submitting price changing
+  # In slave mode this instance confirm changing
+  master: true
```
Comment (antmat) [inline etc/oracle.yaml]: ok for now - but IMO we need automated failover in future

File: insonmnia/oracle/config.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	"time"
+
+	"github.com/jinzhu/configor"
+	"github.com/sonm-io/core/accounts"
+	"github.com/sonm-io/core/blockchain"
+	"github.com/sonm-io/core/insonmnia/logging"
+)
+
+type oracleConfig struct {
+	Mode                 bool          `yaml:"master" default:"false"`
```
Comment (antmat) [inline insonmnia/oracle/config.go]: either use enum, or change naming to something like isMaster (enum is better)

File: insonmnia/oracle/oracle.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	t := util.NewImmediateTicker(o.cfg.Oracle.ContractUpdatePeriod)
+
+	events, err := o.bch.Events().GetEvents(o.ctx, blockchain.GetMultiSigFilter([]common.Address{blockchain.OracleUsdMultisigAddr()}), big.NewInt(0))
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case p := <-priceWatcher.data:
```
Comment (antmat) [inline insonmnia/oracle/oracle.go]: access to channel via Start method not by private member

File: insonmnia/oracle/oracle.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+			if p.err != nil {
+				return p.err
+			}
+			o.actualPrice = p.price
+			o.logger.Debug("Load new price", zap.String("price", p.price.String()))
```
Comment (antmat) [inline insonmnia/oracle/oracle.go]: logs are lowercase, also "loaded" not "load"

File: insonmnia/oracle/oracle.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+			if err != nil {
+				return err
+			}
+			if o.currentPrice.Cmp(currentPrice) != 0 {
+				o.currentPrice = currentPrice
+				o.logger.Debug("Current price changed", zap.String("price", o.currentPrice.String()))
+			}
+			// master mode
+			if o.cfg.Oracle.Mode {
+				go o.SetPrice()
```
Comment (antmat) [inline insonmnia/oracle/oracle.go]: possible goroutine leakage - pass context iniside

File: insonmnia/oracle/oracle.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+			if !ok {
+				return fmt.Errorf("events chanel closed")
+			}
+			// slave mode
+			if !o.cfg.Oracle.Mode {
+				o.logger.Debug("New submission", zap.Any("event", event))
+				switch value := event.Data.(type) {
+				case *blockchain.SubmissionData:
+					if o.checkChanging(value.TransactionId) {
+						go o.confirmChanging(value.TransactionId)
```
Comment (antmat) [inline insonmnia/oracle/oracle.go]: context

File: insonmnia/oracle/oracle.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+	percent.Div(percent, big.NewInt(100))
+
+	if diff.Cmp(percent) == -1 {
+		return true
+	}
+	return false
+}
+
+// checkChanging check transaction parameters
+func (o *Oracle) checkChanging(txID *big.Int) bool {
```
Comment (antmat) [inline insonmnia/oracle/oracle.go]: strange naming
`checkTransaction` or `checkTransactionPrice` or `transactionValid`

File: insonmnia/oracle/oracle.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+// checkPriceChanging check that submitted price does not differ more than 1%
+// incapsulate ugly big math to separate function
+func (o *Oracle) checkPriceChanging(price *big.Int) bool {
+	diff := big.NewInt(0).Set(o.actualPrice)
+	diff.Sub(diff, price)
+	diff.Abs(diff)
+
+	percent := big.NewInt(0).Set(o.actualPrice)
+	percent.Div(percent, big.NewInt(100))
```
Comment (antmat) [inline insonmnia/oracle/oracle.go]: this should be configurable

File: insonmnia/oracle/oracle.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+			zap.String("transactionID", txID.String()),
+			zap.String("transactionTo", transaction.To.String()))
+		return false
+	}
+	d, err := o.bch.OracleUSD().UnpackSetCurrentPriceTransactionData(transaction.Data)
+	if err != nil {
+		return false
+	}
+	if !o.checkPriceChanging(d) {
+		o.logger.Debug("Transaction already executed", zap.String("transactionID", txID.String()))
```
Comment (antmat) [inline insonmnia/oracle/oracle.go]: misleading log

Comment (antmat) [reply insonmnia/oracle/oracle.go]: copy paste?

File: insonmnia/oracle/config.go [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/sonm-io/core/accounts"
+	"github.com/sonm-io/core/blockchain"
+	"github.com/sonm-io/core/insonmnia/logging"
+)
+
+type oracleConfig struct {
+	Mode                 bool          `yaml:"isMaster" default:"false"`
+	PriceUpdatePeriod    time.Duration `yaml:"priceUpdatePeriod" default:"15s"`
+	ContractUpdatePeriod time.Duration `yaml:"contractUpdatePeriod" default:"15m"`
+	Percent              float64       `yaml:"deviationPercent" required:"1.0"`
```
Comment (sshaman1101) [inline insonmnia/oracle/config.go]: I think we should use underscore_case for the YAML annotations.

File: blockchain/api.go [hunk 13]
```diff
@@ -1284,6 +1364,65 @@ func (api *BasicEventsAPI) processLog(log types.Log, eventTS uint64, out chan *E
 			return
 		}
 		sendData(&NumBenchmarksUpdatedData{NumBenchmarks: numBenchmarksBig.Uint64()})
+	case ConfirmationTopic:
+		sender, err := extractAddress(log.Topics, 1)
+		if err != nil {
+			sendErr(out, err, topic)
+			return
+		}
+		transactionId, err := extractBig(log.Topics, 2)
```
Comment (sshaman1101) [inline blockchain/api.go]: here and below: Id -> ID

File: blockchain/api.go [hunk 14]
```diff
@@ -126,6 +129,14 @@ type OracleAPI interface {
 	SetCurrentPrice(ctx context.Context, key *ecdsa.PrivateKey, price *big.Int) (*types.Transaction, error)
 	// GetCurrentPrice returns current price relation between some currency and SONM token
 	GetCurrentPrice(ctx context.Context) (*big.Int, error)
+	// PackSetCurrentPriceTransactionData pack `SetCurrentPrice` method call
+	PackSetCurrentPriceTransactionData(price *big.Int) ([]byte, error)
+	// UnpackSetCurrentPriceTransactionData unpack `SetCurrentPrice` method call
+	UnpackSetCurrentPriceTransactionData(data []byte) (*big.Int, error)
```
Comment (sshaman1101) [inline blockchain/api.go]: I still think that those methods must be encapsulated into the Oracle implementation, furthermore, both of this methods are used once inside the Oracle API implementation itself.

File: insonmnia/oracle/config.go [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+type Config struct {
+	Oracle     oracleConfig       `yaml:"oracle"`
+	Log        logging.Config     `yaml:"log"`
+	Blockchain *blockchain.Config `yaml:"blockchain"`
+	Eth        accounts.EthConfig `yaml:"ethereum" required:"false"`
+}
+
+// NewConfig loads localNode config from given .yaml file
```
Comment (sshaman1101) [inline insonmnia/oracle/config.go]: localNode?

File: insonmnia/oracle/oracle.go [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+	actualPrice  *big.Int
+	currentPrice *big.Int
+}
+
+func NewOracle(ctx context.Context, key *ecdsa.PrivateKey, cfg *Config) (*Oracle, error) {
+	logger := ctxlog.GetLogger(ctx)
+
+	var mode string
+	if cfg.Oracle.Mode {
+		mode = "master"
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: move to constants

File: insonmnia/oracle/oracle.go [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+	} else {
+		mode = "slave"
+	}
+	logger.Info("start USD-SNM Oracle",
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: But you're creating an instance, not starting it. I think this log line should be into the `.Serve()` method.

File: insonmnia/oracle/oracle.go [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+		currentPrice: big.NewInt(0),
+		actualPrice:  big.NewInt(0),
+	}, nil
+}
+
+func (o *Oracle) Serve(ctx context.Context) error {
+	priceWatcher := NewPriceWatcher(o.cfg.Oracle.PriceUpdatePeriod)
+	dw := priceWatcher.Start(ctx)
+
+	t := util.NewImmediateTicker(o.cfg.Oracle.ContractUpdatePeriod)
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: defer ticker stopping

File: insonmnia/oracle/watcher.go [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+func NewPriceWatcher(parsePeriod time.Duration) *PriceWatcher {
+	return &PriceWatcher{
+		data:        make(chan *PriceData),
+		parsePeriod: parsePeriod,
+	}
+}
+
+func (p *PriceWatcher) Start(ctx context.Context) <-chan *PriceData {
+	go func() {
+		t := util.NewImmediateTicker(p.parsePeriod)
```
Comment (sshaman1101) [inline insonmnia/oracle/watcher.go]: defer ticker stopping

File: insonmnia/oracle/oracle.go [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+	events, err := o.bch.Events().GetEvents(ctx, blockchain.GetMultiSigFilter([]common.Address{blockchain.OracleUsdMultisigAddr()}, big.NewInt(0)))
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case p := <-dw:
+			if p.err != nil {
+				return p.err
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: you cannot load price (via remote source using HTTP!) and fail the whole program. IMHO, in this case, you must log error and just retry.

File: insonmnia/oracle/oracle.go [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+			}
+			o.actualPrice = p.price
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: this may be racy, so use mutex

File: insonmnia/oracle/oracle.go [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+			o.logger.Debug("loaded new price", zap.String("price", p.price.String()))
+		case <-t.C:
+			currentPrice, err := o.bch.OracleUSD().GetCurrentPrice(ctx)
+			if err != nil {
+				return err
+			}
+			if o.currentPrice.Cmp(currentPrice) != 0 {
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: what if `<-dw` select file **after** the `t.C`? Price will be empty and everything will be fucked. So explicitly load price for first time, then start timer handling.

File: insonmnia/oracle/config.go [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+	"time"
+
+	"github.com/jinzhu/configor"
+	"github.com/sonm-io/core/accounts"
+	"github.com/sonm-io/core/blockchain"
+	"github.com/sonm-io/core/insonmnia/logging"
+)
+
+type oracleConfig struct {
+	Mode                 bool          `yaml:"isMaster" default:"false"`
```
Comment (sshaman1101) [inline insonmnia/oracle/config.go]: rename the variable as well, `Mode` is really misleading for the bool flag.

File: insonmnia/oracle/oracle.go [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+		case event, ok := <-events:
+			if !ok {
+				return fmt.Errorf("events chanel closed")
+			}
+			// slave mode
+			if !o.cfg.Oracle.Mode {
+				o.logger.Debug("new submission", zap.Any("event", event))
+				switch value := event.Data.(type) {
+				case *blockchain.SubmissionData:
+					if o.transactionValid(ctx, value.TransactionId) {
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: Id -> ID

File: insonmnia/oracle/oracle.go [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+		return err
+	}
+	return o.bch.OracleMultiSig().SubmitTransaction(ctx, o.key, blockchain.OracleUsdAddr(), big.NewInt(0), data)
+}
+
+// checkPrice check that submitted price does not differ more than 1%
+// incapsulate ugly big math to separate function
+func (o *Oracle) checkPrice(price *big.Int) bool {
+	diff := big.NewFloat(0).SetInt(o.actualPrice)
+	diff.Sub(diff, big.NewFloat(0).SetInt(price))
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: why float?

File: insonmnia/oracle/oracle.go [hunk 26]
```diff
@@ -0,0 +1,179 @@
...
+	diff.Abs(diff)
+
+	p := big.NewInt(0).Set(o.actualPrice)
+	p.Div(p, big.NewInt(100))
+	percent := big.NewFloat(o.cfg.Oracle.Percent)
+	percent.Mul(percent, big.NewFloat(0).SetInt(p))
+
+	if diff.Cmp(percent) == -1 {
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: what if you want to increase the price by exactly one percent?

File: insonmnia/oracle/watcher.go [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+	}
+	logger.Debug("Download new price", zap.Float64("price", usdPrice))
+	return p.divideSNM(usdPrice), nil
+}
+
+func (p *PriceWatcher) divideSNM(price float64) *big.Int {
+	return big.NewInt(int64(1 / price * 1e18))
+}
+
+func (p *PriceWatcher) loadSnmPrice(url string) (float64, error) {
```
Comment (sshaman1101) [inline insonmnia/oracle/watcher.go]: please decide once and forever: snm or SNM will be in your code.

File: insonmnia/oracle/watcher.go [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return 0, err
+	}
+	var tickerSnm []*tokenData
+	err = json.Unmarshal(body, &tickerSnm)
+	if err != nil {
+		return 0, err
+	}
+	return strconv.ParseFloat(tickerSnm[0].PriceUsd, 64)
```
Comment (sshaman1101) [inline insonmnia/oracle/watcher.go]: check for slice len before accessing the `tickerSnm[0]`.

File: insonmnia/oracle/oracle.go (hunk 25 above)
Comment (sokel) [reply insonmnia/oracle/oracle.go]: percent must be float type 
e.g. 1.5

File: insonmnia/oracle/oracle.go (hunk 26 above)
Comment (sokel) [reply insonmnia/oracle/oracle.go]: its big.int percent defined in config

File: insonmnia/oracle/oracle.go [hunk 29]
```diff
@@ -0,0 +1,245 @@
...
+	bch          blockchain.API
+	actualPrice  *big.Int
+	currentPrice *big.Int
+	mu           sync.Mutex
+}
+
+func NewOracle(ctx context.Context, key *ecdsa.PrivateKey, cfg *Config) (*Oracle, error) {
+	logger := ctxlog.GetLogger(ctx)
+
+	logger.Info("start USD-SNM Oracle",
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: you already have this log line into `.Serve`, so remove it from here

File: insonmnia/oracle/oracle.go [hunk 30]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+func (o *Oracle) getPriceForSubmit() (*big.Int, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	if o.actualPrice == nil {
+		return nil, fmt.Errorf("actual price is not downloaded")
+	}
+
+	if o.actualPrice.Cmp(big.NewInt(1000000000000000)) < 0 {
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: You can use `params.Finney` from go-ethereum

File: insonmnia/oracle/config.go [hunk 31]
```diff
@@ -0,0 +1,34 @@
...
+	PriceUpdatePeriod    time.Duration `yaml:"price_update_period" default:"15s"`
+	ContractUpdatePeriod time.Duration `yaml:"contract_update_period" default:"15m"`
+	Percent              float64       `yaml:"deviation_percent" required:"1.0"`
+}
+
+type Config struct {
+	Oracle     oracleConfig       `yaml:"oracle"`
+	Log        logging.Config     `yaml:"log"`
+	Blockchain *blockchain.Config `yaml:"blockchain"`
+	Eth        accounts.EthConfig `yaml:"ethereum" required:"false"`
```
Comment (3Hren) [inline insonmnia/oracle/config.go]: A cho required false?

File: insonmnia/oracle/watcher.go [hunk 32]
```diff
@@ -0,0 +1,102 @@
...
+	usdPrice, err := p.loadSNMPrice(snmPriceTickerURL)
+	if err != nil {
+		return nil, err
+	}
+	logger.Debug("Download new price", zap.Float64("price", usdPrice))
+	return p.divideSNM(usdPrice), nil
+}
+
+func (p *PriceWatcher) divideSNM(price float64) *big.Int {
+	return big.NewInt(int64(1 / price * 1e18))
```
Comment (3Hren) [inline insonmnia/oracle/watcher.go]: Starting from `price` ~ 0.1084 and below the result will be negative.

**Pull request merged**